### PR TITLE
E2E for useServerlessTraceTarget

### PR DESCRIPTION
**Pull request by danielcondemarin**
This is literally a copy of `e2e-tests/next-app` with one difference, it sets `useServerlessTraceTarget: true`.

I think we should aim to set serverless trace as the default build target in future, maybe for 1.19. Hopefully the last few fixes and https://github.com/serverless-nextjs/serverless-next.js/pull/779 have made it stable enough.

### CODE REVIEW

File: packages/e2e-tests/next-app-using-serverless-trace/serverless.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+next-app:
```
Comment (dphang) [inline packages/e2e-tests/next-app-using-serverless-trace/serverless.yml:1]: change this name to `next-app-using-serverless-trace

File: packages/e2e-tests/next-app-using-serverless-trace/package.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "next-app",
```
Comment (dphang) [inline packages/e2e-tests/next-app-using-serverless-trace/package.json:2]: Change this to `next-app-using-serverless-trace`

File: packages/e2e-tests/next-app-using-serverless-trace/package.json [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  "version": "1.0.0",
+  "description": "Basic test Next.js app for running serverless-next.js end-to-end tests",
+  "scripts": {
+    "dev": "next",
+    "build": "next build",
+    "start": "next start -p $PORT",
+    "serverless": "serverless",
+    "e2e": "cypress run",
+    "e2e:ci": "SSG_PAGE_PATH=/ssg-page SSR_PAGE_PATH=/ssr-page APP_NAME=next-app ts-node --project tsconfig.scripts.json ../test-utils/scripts/run-e2e-ci.ts"
```
Comment (dphang) [inline packages/e2e-tests/next-app-using-serverless-trace/package.json:11]: Change `next-app` to `next-app-using-serverless-trace`. Needed so it doesn't conflict with the existing e2e tests

**Comment by dphang** [comment]
@danielcondemarin LGTM besides that.

I'm not so sure about setting serverless trace target as the default build though, since although it reduces code size it can have cold start perf impact, since it's now having to load a bunch of smaller JS modules instead of a single file.

There are ways to optimize code size like minifying the pages JS files or manually removing page JS only used for pre-rendering (e.g we do it automatically if you do not use any API routes). I added `postBuildCommands` input which can help with this if you do use API routes.

File: packages/e2e-tests/next-app-using-serverless-trace/package.json (hunk 3 above)
Comment (dphang) [reply packages/e2e-tests/next-app-using-serverless-trace/package.json:11]: Also you need to modify the CI YML file to add this to the strategy matrix

**Review by dphang — APPROVED** [review]

**Comment by danielcondemarin** [comment]
> I'm not so sure about setting serverless trace target as the default build though, since although it reduces code duplication, it can have cold start perf impact, since it's now having to load a bunch of smaller JS modules instead of a single file for a page. From my experience AWS Lambda code size has very little impact on cold start, it's more the I/O/requires that makes up the majority of the time.

That's a fair point. I'm mainly proposing the change because the Next team uses this as their build target for Vercel deployments. The other is that I suspect they will be deprecating the serverless target. 
Happy to keep as as an opt-in for now. I might do a bit more digging on this area and post any findings.

**Comment by dphang** [comment]
> > I'm not so sure about setting serverless trace target as the default build though, since although it reduces code duplication, it can have cold start perf impact, since it's now having to load a bunch of smaller JS modules instead of a single file for a page. From my experience AWS Lambda code size has very little impact on cold start, it's more the I/O/requires that makes up the majority of the time.
> 
> That's a fair point. I'm mainly proposing the change because the Next team uses this as their build target for Vercel deployments. The other is that I suspect they will be deprecating the serverless target.
> Happy to keep serverless-trace as an opt-in for now. I might do a bit more digging on this area and post any findings.

Yep, I did try Vercel's deployment and found it slow for that same reason - a while back I had opened an issue but had no responses. I hope they do not remove the serverless target.

**Comment by dphang** [comment]
Looks like e2e tests failed here: https://github.com/serverless-nextjs/serverless-next.js/actions/runs/358701235

File: packages/e2e-tests/next-app-using-serverless-trace/next.config.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+module.exports = function (...args) {
```
Comment (dphang) [inline packages/e2e-tests/next-app-using-serverless-trace/next.config.js:1]: Woops, overlooked this, I will fix for you @danielcondemarin

**Pull request merged**